metagraph/src/cli/align.cpp

```diff
@@ -426,21 +414,29 @@ int align_to_graph(Config *config) {
             }
 
             auto process_batch = [&,graph](SeqBatch batch) {
-                // make a shared_ptr in a thread-safe way
-                std::shared_ptr<DeBruijnGraph> aln_graph(
-                    std::shared_ptr<DeBruijnGraph>{}, graph.get()
-                );
+                // make an aliased shared_ptr in a thread-safe way
+                auto aln_graph = std::shared_ptr<DeBruijnGraph>(std::shared_ptr<DeBruijnGraph>{}, graph.get());
+
+                if (aln_graph->get_mode() == DeBruijnGraph::PRIMARY)
+                    aln_graph = std::make_shared<CanonicalDBG>(aln_graph, 100'000);
 
-                if (auto *canonical = dynamic_cast<CanonicalDBG*>(aln_graph.get()))
-                    aln_graph = std::make_shared<CanonicalDBG>(*canonical);
+                std::shared_ptr<AnnotatedDBG> aln_anno_graph;
+                std::shared_ptr<IDBGAligner> aligner;
 
-                auto aligner = build_aligner(*aln_graph, aligner_config);
+                if (annotator) {
+                    aln_anno_graph = std::make_shared<AnnotatedDBG>(aln_graph, annotator);
```

Comment:
You're creating an annotated graph from a pair of potentially incompatible (graph, annotator) pair.

The annotator is created for `base_graph` and hence should be used with it.
Now you wrap `base_graph` into something else (`aln_graph`), so it's no longer compatible.
You are not supposed to create such cases.

I guess this is why you need `get_base_node` and stuff in downstream to recover access to the graph with which you use the aligner?

Reply:
It depends on how "compatibility" with AnnotatedDBG is defined. If it's based on simply the number of nodes in the graph, then yes, this breaks things.

However, if you base compatibility on the functionality of the following workflow:
1) run `map_to_nodes` on the sequence
2) extract labels for the returned nodes.

Then compatibility shouldn't be broken.

Reply:
The place where the inconsistency happens is that other operations like `map_to_nodes_sequentially` and `call_outgoing_kmers` do not map to/return canonical k-mers (because this would lead to false traversal cycles). So for these, there needs to be some interface for fetching labels on these calls.

Reply:
It's already defined by `check_compatibility()`. For some reason it was missing in the constructor, so your tests passed. As you see, they don't anymore: 8e0adfd

In fact, this was checked in any method. So you just created that class and never called any method from it.

Reply:
`check_compatibility` is defined differently between `AnnotatedSequenceGraph` and `AnnotatedDBG`. In the latter, it checks the base graph, rather than the graph itself.

However, if something like `get_base_graph` existed in `SequenceGraph`, then they could be defined the same way.

Reply:
In the end, I didn't use the `AnnotatedDBG` interface because it would have added a bunch of overhead to the workflow.

Essentially, I need something that can take a bunch of graph nodes (which may or may not be from the wrapper), and map fetch a vector of label IDs for each one. I want to be able to do this in a way that minimizes remapping to the graph, and also doesn't involve a bunch of unnecessary conversion between labels and label IDs.

Reply:
I don't see a problem with them being defined differently.
What does it have to do with breaking the compatibility in this line?

Reply:
The check was changed in `9f4e8e5b73571d874dd93eedc53079f71615023c`. So now it's not even possible to add check_compatibility() to the constructor...

On the other hand, without that check, everything is compatible. Brilliant.